### PR TITLE
Add a way to specify username to send notifications to

### LINKED ISSUE #14748: Remove useless `NotifyUserTask`
I think the `NotifyUserTask` is useless and should be removed. 

Unless we salvage it by adding a required username input so the user can specify a username to notify.

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #14748

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: This is almost entirely the same as `NotifyContentOwnerTask.Fields.Edit` and `NotifyUserTaskActivity.Fields.Edit`. Can't we DRY these? And actually, so we need the latter one, to begin with?

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs:44]: How about having some validation here that prevents you from using more than x user names? There's a limit for `IN` expressions, [at least with SQL Server](https://stackoverflow.com/questions/21178390/in-clause-limitation-in-sql-server).

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs:35]:
```suggestion
        var userBames = viewModel.Usernames?.Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries);
```

**Review by Piedone — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs:39]: @MikeAlhayek 
I rather keep it as `username` UserName to me is the name of the user like "Mike Alhayek". Username is an identifier (for example, malhayek) so it would be Username.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: we probably can create multiple views one would render the base and one would render only the properties we need. More work than I want to do in this PR :)

File: src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs:39]: The casing is `UserName` for this identifier otherwise, including in `UserIndex` where it references it.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: I wouldn't ask about this if this PR wouldn't add another copy.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs:46]: sure. I added this.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: This works for me. Make sure you pull the latest commit from this branch. At some point it was not working but it does now. maybe cached dll too. `npm run cleanup`

**Comment by Piedone** [comment]
Please only request review if we have all the discussions settled. I really don't want to keep getting back to things.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs:39]: If you search our code you'll see that Username is used in most cases not UserName. From a language perspective, I think the correct test for a unique identifier is `username` where `UserName` (User Name) referees to the user's name

File: src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: What's in the video happens for me even after a `git clean` and pulling the latest commit (`041e1bde285278ba45ca34a7c8c1f3ce2519a122`). Latest Chrome Guest profile (i.e. stock options) under Windows. Same with a latest FF private window without any extensions.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: Not sure. Are there any exceptions in your log file? It works for me with no problems. Here is a proof

user-activity

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: Okay. I made the change so anyone that in inheriting from `NotifyUserTaskActivityDisplayDriver` does not have to worry about the base fields/view.

**Comment by MikeAlhayek** [comment]
@Piedone I made the requested changes. I would rather keep `Username` instead of `UserName`. Let me know if you have any additional changes.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs:39]: The only place where the name of the user account is not called UserName in members is in the `UsernameOrEmail` members of the forgot and reset password features:

image

"UserName" turns up 322 matches while "userName" 119. It's "UserName" as the property of `User` too, what all of the changes in this PR refer to.

This includes `UserIndex` what this PR uses, so there's "_session.Query<User, UserIndex>(user => user.NormalizedUser**N**ame.IsIn(user**n**ames))."

We only have this kind of name, i.e. the unique identifier, in the system. While we could have a linguistic debate if this were the start of the project, and I don't have a strong feeling either way, it is definitely "UserName" currently.

Let's just keep the consistency.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: Note that this isn't exactly my repro: You also added an Html Body and ticked "Use HTML Body when possible". If I do that (actually only adding something to the Body is necessary) then it works for me too.

I don't have any exceptions in the log.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: Great!

File: src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs]: I guess if we are wrong, let's just be wrong "almost" everywhere. I made this change

File: src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Activities/NotifyUserTask.cs:51]: Done.

File: src/OrchardCore.Modules/OrchardCore.Notifications/ViewModels/NotifyUserTaskViewModel.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/ViewModels/NotifyUserTaskViewModel.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: The Liquid validator was throwing null reference exception. You uncovered existing bug in main. Good job

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Views/Items/NotifyUserTask.Fields.Edit.cshtml:1]: Good then :). Interesting that it's not logged though.

**Comment by MikeAlhayek** [comment]
@Piedone I am done with this. I made some changes to simplify inheriting the base display driver. Let me know if you see anything else. Otherwise, please approve. I would like to test this out using "preview" in my project tomorrow.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (Piedone) [inline src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs:45]: For readability in the text box:

```suggestion
            activity.UserNames = new WorkflowExpression<string>(string.Join(", ", userNames));
```

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: I don't see this being addressed.

**Review by Piedone — CHANGES_REQUESTED** [review]
You can BTW `dotnet pack` just this project locally with a higher version number, reference it from your web project with a local package source, and test it right away.

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: done

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: I still don't see this being addressed.

**Comment by MikeAlhayek** [comment]
> You can BTW dotnet pack just this project locally with a higher version number, reference it from your web project with a local package source, and test it right away.

@Piedone this would require you to pack 4 different packages since more than one project was updated. I will wait until this is merged since everything was addressed anyway and the PR is ready.

**Comment by Piedone** [comment]
Not everything: https://github.com/OrchardCMS/OrchardCore/pull/16073#discussion_r1603808601

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: I added a distinct check to reduce the number of values. We can't really write a statement to say you can only have 2 X amount of parameters because the value can be different from one server to another. It's the same reason why in the content manager we allow you to provide as many contentItemIds as you need. We don't validate at that level. Adding Distinct will help.

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: OK, let's leave it like that.

**Review by Piedone — APPROVED** [review]

File: src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: @Piedone I just made a change to query the users in chunks of 1000 per request. Happy? :)

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: Did you push it too?

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: https://github.com/OrchardCMS/OrchardCore/pull/16073/commits/acd9569abd3d0676289dd0edfd8dd52e7816a5d7

Comment (Piedone) [reply src/OrchardCore.Modules/OrchardCore.Notifications/Drivers/NotifyUserTaskDisplayDriver.cs]: Great!

**Pull request merged**
